### ISSUE TITLE
kubeadm timeout updating kube-scheduler static pod

### ISSUE THREAD

**Issue by wToad** [kubernetes/kubeadm#1603]
## What keywords did you search in kubeadm issues before filing this one?

```
kubeadm kube-scheduler couldn't upgrade control plane
```

## Versions
- Kubeadm version:
```
kubeadm version: &version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.7", GitCommit:"6f482974b76db3f1e0f5d24605a9d1d38fad9a2b", GitTreeState:"clean", BuildDate:"2019-03-25T02:49:02Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
```
- Kubernetes version:
```
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.7", GitCommit:"6f482974b76db3f1e0f5d24605a9d1d38fad9a2b", GitTreeState:"clean", BuildDate:"2019-03-25T02:41:57Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.3", GitCommit:"a4529464e4629c21224b3d52edfe0ea91b072862", GitTreeState:"clean", BuildDate:"2018-09-09T17:53:03Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider:
```
Openstack
```
- OS:
```
# lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.5 LTS
Release:	16.04
Codename:	xenial
```
- Kernel:
```
Linux preprod-k8s-master0 4.4.0-150-generic #176-Ubuntu SMP Wed May 29 18:56:26 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```
## What happened?

I wanted to upgrade my cluster with Kubespray, but some errors occured. I decided to do it with kubeadm (even if kubeadm was disable on the kubespray cluster deployment).

I use the simplest upgrade commands:
```
# kubeadm upgrade plan v1.12.7
[preflight] Running pre-flight checks.
[upgrade] Making sure the cluster is healthy:
[upgrade/config] Making sure the configuration is correct:
[upgrade/config] Reading configuration from the cluster...
[upgrade/config] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
[upgrade] Fetching available versions to upgrade to
[upgrade/versions] Cluster version: v1.11.3
[upgrade/versions] kubeadm version: v1.12.7

Upgrade to the latest version in the v1.11 series:

COMPONENT            CURRENT   AVAILABLE
API Server           v1.11.3   v1.12.7
Controller Manager   v1.11.3   v1.12.7
Scheduler            v1.11.3   v1.12.7
Kube Proxy           v1.11.3   v1.12.7
CoreDNS                        1.2.2

You can now apply the upgrade by executing the following command:

	kubeadm upgrade apply v1.12.7

_____________________________________________________________________

# kubeadm upgrade apply v1.12.7
[preflight] Running pre-flight checks.
[upgrade] Making sure the cluster is healthy:
[upgrade/config] Making sure the configuration is correct:
[upgrade/config] Reading configuration from the cluster...
[upgrade/config] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
[upgrade/apply] Respecting the --cri-socket flag that is set with higher priority than the config file.
[upgrade/version] You have chosen to change the cluster version to "v1.12.7"
[upgrade/versions] Cluster version: v1.11.3
[upgrade/versions] kubeadm version: v1.12.7
[upgrade/confirm] Are you sure you want to proceed with the upgrade? [y/N]: Y
[upgrade/prepull] Will prepull images for components [kube-apiserver kube-controller-manager kube-scheduler etcd]
[upgrade/prepull] Prepulling image for component etcd.
[upgrade/prepull] Prepulling image for component kube-apiserver.
[upgrade/prepull] Prepulling image for component kube-controller-manager.
[upgrade/prepull] Prepulling image for component kube-scheduler.
[apiclient] Found 0 Pods for label selector k8s-app=upgrade-prepull-kube-apiserver
[apiclient] Found 0 Pods for label selector k8s-app=upgrade-prepull-etcd
[apiclient] Found 0 Pods for label selector k8s-app=upgrade-prepull-kube-controller-manager
[apiclient] Found 0 Pods for label selector k8s-app=upgrade-prepull-kube-scheduler
[upgrade/prepull] Failed prepulled the images for the control plane components error: unable to cleanup the DaemonSet used for prepulling etcd: daemonsets.apps "upgrade-prepull-etcd" not found
root@preprod-k8s-master0:~# kubeadm upgrade apply v1.12.7
[preflight] Running pre-flight checks.
[upgrade] Making sure the cluster is healthy:
[upgrade/config] Making sure the configuration is correct:
[upgrade/config] Reading configuration from the cluster...
[upgrade/config] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
[upgrade/apply] Respecting the --cri-socket flag that is set with higher priority than the config file.
[upgrade/version] You have chosen to change the cluster version to "v1.12.7"
[upgrade/versions] Cluster version: v1.11.3
[upgrade/versions] kubeadm version: v1.12.7
[upgrade/confirm] Are you sure you want to proceed with the upgrade? [y/N]: Y
[upgrade/prepull] Will prepull images for components [kube-apiserver kube-controller-manager kube-scheduler etcd]
[upgrade/prepull] Prepulling image for component etcd.
[upgrade/prepull] Prepulling image for component kube-apiserver.
[upgrade/prepull] Prepulling image for component kube-controller-manager.
[upgrade/prepull] Prepulling image for component kube-scheduler.
[apiclient] Found 0 Pods for label selector k8s-app=upgrade-prepull-kube-controller-manager
[apiclient] Found 0 Pods for label selector k8s-app=upgrade-prepull-kube-scheduler
[apiclient] Found 0 Pods for label selector k8s-app=upgrade-prepull-kube-apiserver
[apiclient] Found 0 Pods for label selector k8s-app=upgrade-prepull-etcd
[upgrade/prepull] Prepulled image for component kube-apiserver.
[upgrade/prepull] Prepulled image for component kube-controller-manager.
[upgrade/prepull] Prepulled image for component kube-scheduler.
[upgrade/prepull] Prepulled image for component etcd.
[upgrade/prepull] Successfully prepulled the images for all the control plane components
[upgrade/apply] Upgrading your Static Pod-hosted control plane to version "v1.12.7"...
Static pod: kube-apiserver-openpaas-preprod-linagora-k8s-master0 hash: cacaf59666c7f9ad5fd5cfb34119fd91
Static pod: kube-controller-manager-openpaas-preprod-linagora-k8s-master0 hash: dd30aded3667c0791198bced69e54dd0
Static pod: kube-scheduler-openpaas-preprod-linagora-k8s-master0 hash: b8848d630ad36ddc239268433b5e1efa
[upgrade/staticpods] Writing new Static Pod manifests to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests255994404"
[controlplane] wrote Static Pod manifest for component kube-apiserver to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests255994404/kube-apiserver.yaml"
[controlplane] wrote Static Pod manifest for component kube-controller-manager to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests255994404/kube-controller-manager.yaml"
[controlplane] wrote Static Pod manifest for component kube-scheduler to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests255994404/kube-scheduler.yaml"
[upgrade/staticpods] Moved new manifest to "/etc/kubernetes/manifests/kube-apiserver.yaml" and backed up old manifest to "/etc/kubernetes/tmp/kubeadm-backup-manifests-2019-06-11-10-23-23/kube-apiserver.yaml"
[upgrade/staticpods] Waiting for the kubelet to restart the component
[upgrade/staticpods] This might take a minute or longer depending on the component/version gap (timeout 5m0s
Static pod: kube-apiserver-openpaas-preprod-linagora-k8s-master0 hash: 72a52cbf43953e22118f505819134075
[apiclient] Found 1 Pods for label selector component=kube-apiserver
[upgrade/staticpods] Component "kube-apiserver" upgraded successfully!
[upgrade/staticpods] Moved new manifest to "/etc/kubernetes/manifests/kube-controller-manager.yaml" and backed up old manifest to "/etc/kubernetes/tmp/kubeadm-backup-manifests-2019-06-11-10-23-23/kube-controller-manager.yaml"
[upgrade/staticpods] Waiting for the kubelet to restart the component
[upgrade/staticpods] This might take a minute or longer depending on the component/version gap (timeout 5m0s
Static pod: kube-controller-manager-openpaas-preprod-linagora-k8s-master0 hash: dd30aded3667c0791198bced69e54dd0
Static pod: kube-controller-manager-openpaas-preprod-linagora-k8s-master0 hash: 6dee512b0920e93ade6a32273e43fa63
[apiclient] Found 1 Pods for label selector component=kube-controller-manager
[upgrade/staticpods] Component "kube-controller-manager" upgraded successfully!
[upgrade/staticpods] Moved new manifest to "/etc/kubernetes/manifests/kube-scheduler.yaml" and backed up old manifest to "/etc/kubernetes/tmp/kubeadm-backup-manifests-2019-06-11-10-23-23/kube-scheduler.yaml"
[upgrade/staticpods] Waiting for the kubelet to restart the component
[upgrade/staticpods] This might take a minute or longer depending on the component/version gap (timeout 5m0s
Static pod: kube-scheduler-openpaas-preprod-linagora-k8s-master0 hash: b8848d630ad36ddc239268433b5e1efa
Static pod: kube-scheduler-openpaas-preprod-linagora-k8s-master0 hash: b8848d630ad36ddc239268433b5e1efa
Static pod: kube-scheduler-openpaas-preprod-linagora-k8s-master0 hash: b8848d630ad36ddc239268433b5e1efa
Static pod: kube-scheduler-openpaas-preprod-linagora-k8s-master0 hash: b8848d630ad36ddc239268433b5e1efa
[...]
[upgrade/apply] FATAL: couldn't upgrade control plane. kubeadm has tried to recover everything into the earlier state. Errors faced: [timed out waiting for the condition]
```

It seems that it is stuck comparing the same pod hash over and over. A [similar issue](https://github.com/kubernetes/kubeadm/issues/1546) as been posted but I could not reopen it. Someone there asked for the manifest, here is it:
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    scheduler.alpha.kubernetes.io/critical-pod: ""
  creationTimestamp: null
  labels:
    component: kube-scheduler
    tier: control-plane
  name: kube-scheduler
  namespace: kube-system
spec:
  containers:
  - command:
    - kube-scheduler
    - --feature-gates=PersistentLocalVolumes=False,VolumeScheduling=False,MountPropagation=False
    - --profiling=False
    - --address=127.0.0.1
    - --kubeconfig=/etc/kubernetes/scheduler.conf
    - --leader-elect=true
    image: gcr.io/google-containers/hyperkube-amd64:v1.11.3
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 8
      httpGet:
        host: 127.0.0.1
        path: /healthz
        port: 10251
        scheme: HTTP
      initialDelaySeconds: 15
      timeoutSeconds: 15
    name: kube-scheduler
    resources:
      requests:
        cpu: 100m
    volumeMounts:
    - mountPath: /etc/kubernetes/scheduler.conf
      name: kubeconfig
      readOnly: true
  hostNetwork: true
  priorityClassName: system-cluster-critical
  volumes:
  - hostPath:
      path: /etc/kubernetes/scheduler.conf
      type: FileOrCreate
    name: kubeconfig
status: {}
```

**Comment by neolit123** [comment]
you mentioned in the other issue (https://github.com/kubernetes/kubeadm/issues/1546) that you hand problems (possibly corruption) with the cluster CA.

could you please copy / add the same comments here?

on the side of kubeadm we have a problem that our errors are not good - i.e. it's not clear what the problem is.
but it feels to me that we have a case of a unsupported cluster where something is not right with it's certificates.

sharing again that our 1.11 -> 1.12 upgrades work:
https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-upgrade-1.11-1.12

**Comment by wToad** [comment]
sure, so I checked the logs of the container:
```
E0611 15:04:01.967918       1 reflector.go:205] k8s.io/kubernetes/vendor/k8s.io/client-go/informers/factory.go:130: Failed to list *v1.Node: Get https://51.68.112.134:6443/api/v1/nodes?limit=500&resourceVersion=0: x509: certificate signed by unknown authority
```
There are a lot of errors mentionning this certificate issue. I had many troubles using Kubespray before kubeadm, and since that, I met few issues according CA. So I think because of Kubespray I should manage it now

I'm still stuck on this but I'm not working on it right now, I'm having a break and do other stuff so I don't have more clue right now

**Comment by neolit123** [comment]
hi, did you resolve your problem?

> 1 reflector.go:205] k8s.io/kubernetes/vendor/k8s.io/client-go/informers/factory.go:130: Failed to list *v1.Node: Get https://51.68.112.134:6443/api/v1/nodes?limit=500&resourceVersion=0: x509: certificate signed by unknown authority

this means that the client that attempts to perform an operation against the cluster is singed with a different CA. i don't know why this is happening but you should ask the kubespray maintainers at this point.

i was investigating something else and saw this in your logs:
> [upgrade/prepull] Failed prepulled the images for the control plane components error: unable to cleanup the DaemonSet used for prepulling etcd: daemonsets.apps "upgrade-prepull-etcd" not found

see:
https://github.com/kubernetes/kubeadm/issues/1700

a patch for that might be:
https://github.com/kubernetes/kubernetes/pull/80798
but it will be only available in 1.16.

**Comment by neolit123** [comment]
please reopen if needed.
